File: app/src/main/java/be/digitalia/fosdem/widgets/SaveStatePhotoView.java
package be.digitalia.fosdem.widgets;

import android.content.Context;
import android.graphics.RectF;
import android.os.Parcel;
import android.os.Parcelable;
import android.util.AttributeSet;
import android.view.ViewTreeObserver;
import com.github.chrisbanes.photoview.PhotoView;

/**
 * PhotoView which saves and restores the current scale and approximate position.
 */
public class SaveStatePhotoView extends PhotoView {

	public SaveStatePhotoView(Context context) {
		super(context);
	}

	public SaveStatePhotoView(Context context, AttributeSet attr) {
		super(context, attr);
	}

	public SaveStatePhotoView(Context context, AttributeSet attr, int defStyle) {
		super(context, attr, defStyle);
	}

	@Override
	protected Parcelable onSaveInstanceState() {
		Parcelable superState = super.onSaveInstanceState();
		SavedState ss = new SavedState(superState);
		ss.scale = getScale();
		final RectF rect = getDisplayRect();
		final float overflowWidth = rect.width() - getWidth();
		if (overflowWidth > 0f) {
			ss.pivotX = -rect.left / overflowWidth;
		}
		final float overflowHeight = rect.height() - getHeight();
		if (overflowHeight > 0f) {
			ss.pivotY = -rect.top / overflowHeight;
		}
		return ss;
	}

	@Override
	protected void onRestoreInstanceState(Parcelable state) {
		if (!(state instanceof SavedState)) {
			super.onRestoreInstanceState(state);
			return;
		}

		final SavedState ss = (SavedState) state;
		super.onRestoreInstanceState(ss.getSuperState());

		getViewTreeObserver().addOnGlobalLayoutListener(new ViewTreeObserver.OnGlobalLayoutListener() {
			@Override
			public void onGlobalLayout() {
				float scale = Math.max(ss.scale, getMinimumScale());
				scale = Math.min(scale, getMaximumScale());
				setScale(scale, getWidth() * ss.pivotX, getHeight() * ss.pivotY, false);
				getViewTreeObserver().removeOnGlobalLayoutListener(this);
			}
		});
	}

	public static class SavedState extends BaseSavedState {
		float scale = 1f;
		float pivotX = 0.5f;
		float pivotY = 0.5f;

		public SavedState(Parcelable superState) {
			super(superState);
		}

		@Override
		public void writeToParcel(Parcel out, int flags) {
			super.writeToParcel(out, flags);
			out.writeFloat(scale);
			out.writeFloat(pivotX);
			out.writeFloat(pivotY);
		}

		public static final Parcelable.Creator<SavedState> CREATOR = new Parcelable.Creator<SavedState>() {

			@Override
			public SavedState createFromParcel(Parcel in) {
				return new SavedState(in);
			}

			@Override
			public SavedState[] newArray(int size) {
				return new SavedState[size];
			}
		};

		SavedState(Parcel in) {
			super(in);
			scale = in.readFloat();
			pivotX = in.readFloat();
			pivotY = in.readFloat();
		}
	}
}

File: app/src/main/java/be/digitalia/fosdem/widgets/SaveStatePhotoView.kt
package be.digitalia.fosdem.widgets

import android.content.Context
import android.os.Parcel
import android.os.Parcelable
import android.util.AttributeSet
import androidx.core.view.doOnLayout
import androidx.customview.view.AbsSavedState
import com.github.chrisbanes.photoview.PhotoView

/**
 * PhotoView which saves and restores the current scale and approximate position.
 */
class SaveStatePhotoView : PhotoView {

    constructor(context: Context) : super(context)
    constructor(context: Context, attr: AttributeSet?) : super(context, attr)
    constructor(context: Context, attr: AttributeSet?, defStyle: Int) : super(context, attr, defStyle)

    override fun onSaveInstanceState(): Parcelable? {
        val superState = super.onSaveInstanceState()
        val rect = displayRect
        val overflowWidth = rect.width() - width
        val pivotX = if (overflowWidth > 0f) {
            -rect.left / overflowWidth
        } else 0.5f
        val overflowHeight = rect.height() - height
        val pivotY = if (overflowHeight > 0f) {
            -rect.top / overflowHeight
        } else 0.5f
        return SavedState(superState ?: AbsSavedState.EMPTY_STATE, scale, pivotX, pivotY)
    }

    override fun onRestoreInstanceState(state: Parcelable) {
        if (state !is SavedState) {
            super.onRestoreInstanceState(state)
            return
        }

        super.onRestoreInstanceState(state.superState)

        doOnLayout {
            setScale(state.scale.coerceIn(minimumScale, maximumScale),
                    width * state.pivotX,
                    height * state.pivotY,
                    false)
        }
    }

    class SavedState : AbsSavedState {

        val scale: Float
        val pivotX: Float
        val pivotY: Float

        constructor(superState: Parcelable, scale: Float, pivotX: Float, pivotY: Float) : super(superState) {
            this.scale = scale
            this.pivotX = pivotX
            this.pivotY = pivotY
        }

        override fun writeToParcel(out: Parcel, flags: Int) {
            super.writeToParcel(out, flags)
            out.writeFloat(scale)
            out.writeFloat(pivotX)
            out.writeFloat(pivotY)
        }

        private constructor(source: Parcel, loader: ClassLoader?) : super(source, loader) {
            scale = source.readFloat()
            pivotX = source.readFloat()
            pivotY = source.readFloat()
        }

        companion object {
            @JvmField
            @Suppress("UNUSED")
            val CREATOR = object : Parcelable.ClassLoaderCreator<SavedState> {
                override fun createFromParcel(source: Parcel, loader: ClassLoader?) = SavedState(source, loader)

                override fun createFromParcel(source: Parcel) = SavedState(source, null)

                override fun newArray(size: Int) = arrayOfNulls<SavedState>(size)
            }
        }
    }
}